packages/flutter/test/material/tabs_test.dart

```diff
@@ -2580,4 +2580,81 @@ void main() {
     await tester.pumpAndSettle();
     pageController.removeListener(pageControllerListener);
   });
+
+  testWidgets('onTabChanged responds to Tab click and Scroll Notifications', (WidgetTester tester) async {
+    const List<Tab> tabs = <Tab>[
+      Tab(text: 'Zeroth'), Tab(text: 'First'), Tab(text: 'Second')
+    ];
+    final List<int> logs = <int>[];
+
+    await tester.pumpWidget(MaterialApp(
+        home: DefaultTabController(
+          length: tabs.length,
+          child: Scaffold(
+            appBar: AppBar(
+              bottom: const TabBar(tabs: tabs),
+            ),
+            body: TabBarView(
+              children: tabs.map((Tab tab) => Center(child: Text('${tab.text} Tab'))).toList(),
+              onTabChanged: (int tabNumber) {
+                logs.add(tabNumber);
+              },
+            ),
+          ),
+        )));
+
+    // Initially logs should be empty.
+    expect(logs.isEmpty, isTrue);
+
+    // Change the Tab through TabBar.
+    await tester.tap(find.text('Second'));
+    await tester.pumpAndSettle();
+
+    expect(logs.first, 2);
+    logs.removeLast();
+
+    // Change the Tab through swipe gesture.
+    await tester.fling(find.text('Second Tab'), const Offset(400, 0), 800);
+    await tester.pumpAndSettle();
+
+    expect(logs.first, 1);
+  });
+
+  testWidgets('onTabChanged is called only when tab is changed', (WidgetTester tester) async {
+    const List<Tab> tabs = <Tab>[
+      Tab(text: 'Zeroth'), Tab(text: 'First'), Tab(text: 'Second')
+    ];
+    final List<int> logs = <int>[];
+
+    // This test indicates that, even if user is scrolling, the onTabChanged is
+    // called only when current page is changed.
+    await tester.pumpWidget(MaterialApp(
+        home: DefaultTabController(
+          length: tabs.length,
+          child: Scaffold(
+            appBar: AppBar(
+              bottom: const TabBar(tabs: tabs),
+            ),
+            body: TabBarView(
+              children: tabs.map((Tab tab) => Center(child: Text('${tab.text} Tab'))).toList(),
+              onTabChanged: (int tabNumber) {
+                logs.add(tabNumber);
+              },
+            ),
+          ),
+        )));
```

Comment:
same comments as above:
```suggestion
    await tester.pumpWidget(MaterialApp(
      home: DefaultTabController(
        length: tabs.length,
        child: Scaffold(
          appBar: AppBar(
            bottom: const TabBar(tabs: tabs),
          ),
          body: TabBarView(
            onTabChanged: (int tabNumber) {
              logs.add(tabNumber);
            },
            children: tabs.map((Tab tab) => Center(child: Text('${tab.text} Tab'))).toList(),            
          ),
        ),
      ),
    ));
```